File: server/src/PhotoSync.Core/Entities/Photo.cs
namespace PhotoSync.Core.Entities;

/// <summary>
/// Represents a synced photo stored on the server.
/// </summary>
public sealed class Photo
{
    /// <summary>
    /// Unique identifier for the photo.
    /// </summary>
    public Guid Id { get; private set; }

    /// <summary>
    /// Original filename from the device.
    /// </summary>
    public string OriginalFilename { get; private set; } = string.Empty;

    /// <summary>
    /// Relative path where the file is stored (e.g., "2024/03/IMG_xxx.jpg").
    /// </summary>
    public string StoredPath { get; private set; } = string.Empty;

    /// <summary>
    /// SHA256 hash of the file content for duplicate detection.
    /// </summary>
    public string FileHash { get; private set; } = string.Empty;

    /// <summary>
    /// Size of the file in bytes.
    /// </summary>
    public long FileSize { get; private set; }

    /// <summary>
    /// Date the photo was taken (from device metadata).
    /// </summary>
    public DateTime DateTaken { get; private set; }

    /// <summary>
    /// Date/time when the photo was uploaded to the server.
    /// </summary>
    public DateTime UploadedAt { get; private set; }

    // Private constructor for EF Core
    private Photo() { }

    /// <summary>
    /// Creates a new Photo entity.
    /// </summary>
    public static Photo Create(
        string originalFilename,
        string storedPath,
        string fileHash,
        long fileSize,
        DateTime dateTaken)
    {
        if (string.IsNullOrWhiteSpace(originalFilename))
            throw new ArgumentException("Original filename cannot be empty.", nameof(originalFilename));

        if (string.IsNullOrWhiteSpace(storedPath))
            throw new ArgumentException("Stored path cannot be empty.", nameof(storedPath));

        if (string.IsNullOrWhiteSpace(fileHash))
            throw new ArgumentException("File hash cannot be empty.", nameof(fileHash));

        if (fileSize <= 0)
            throw new ArgumentException("File size must be positive.", nameof(fileSize));

        return new Photo
        {
            Id = Guid.NewGuid(),
            OriginalFilename = SanitizeFilename(originalFilename),
            StoredPath = storedPath,
            FileHash = fileHash.ToLowerInvariant(),
            FileSize = fileSize,
            DateTaken = dateTaken,
            UploadedAt = DateTime.UtcNow
        };
    }

    /// <summary>
    /// Sanitizes a filename to prevent path traversal attacks.
    /// </summary>
    private static string SanitizeFilename(string filename)
    {
        // Remove any path components - only keep the filename
        var name = Path.GetFileName(filename);

        // Remove any potentially dangerous characters
        var invalidChars = Path.GetInvalidFileNameChars();
        foreach (var c in invalidChars)
        {
            name = name.Replace(c, '_');
        }

        return name;
    }
}

File: server-go/internal/models/photo.go
package models

import (
	"path/filepath"
	"strings"
	"time"

	"github.com/google/uuid"
)

// Photo represents a synced photo stored on the server
type Photo struct {
	ID               string    `json:"id"`
	OriginalFilename string    `json:"originalFilename"`
	StoredPath       string    `json:"storedPath"`
	FileHash         string    `json:"fileHash"`
	FileSize         int64     `json:"fileSize"`
	DateTaken        time.Time `json:"dateTaken"`
	UploadedAt       time.Time `json:"uploadedAt"`
	UserID           *string   `json:"userId,omitempty"`
	OriginDeviceID   *string   `json:"originDeviceId,omitempty"`

	// Thumbnail paths (relative to storage base)
	ThumbSmall  *string `json:"thumbSmall,omitempty"`
	ThumbMedium *string `json:"thumbMedium,omitempty"`
	ThumbLarge  *string `json:"thumbLarge,omitempty"`

	// EXIF Metadata
	CameraMake   *string `json:"cameraMake,omitempty"`
	CameraModel  *string `json:"cameraModel,omitempty"`
	LensModel    *string `json:"lensModel,omitempty"`
	FocalLength  *string `json:"focalLength,omitempty"`
	Aperture     *string `json:"aperture,omitempty"`
	ShutterSpeed *string `json:"shutterSpeed,omitempty"`
	ISO          *int    `json:"iso,omitempty"`
	Orientation  int     `json:"orientation"`

	// GPS Location
	Latitude  *float64 `json:"latitude,omitempty"`
	Longitude *float64 `json:"longitude,omitempty"`
	Altitude  *float64 `json:"altitude,omitempty"`

	// Image dimensions
	Width  *int `json:"width,omitempty"`
	Height *int `json:"height,omitempty"`
}

// NewPhoto creates a new Photo with validation and sanitization
func NewPhoto(originalFilename, storedPath, fileHash string, fileSize int64, dateTaken time.Time) (*Photo, error) {
	if strings.TrimSpace(originalFilename) == "" {
		return nil, ErrEmptyFilename
	}
	if strings.TrimSpace(storedPath) == "" {
		return nil, ErrEmptyStoredPath
	}
	if strings.TrimSpace(fileHash) == "" {
		return nil, ErrEmptyHash
	}
	if fileSize <= 0 {
		return nil, ErrInvalidFileSize
	}

	return &Photo{
		ID:               uuid.New().String(),
		OriginalFilename: sanitizeFilename(originalFilename),
		StoredPath:       storedPath,
		FileHash:         strings.ToLower(fileHash),
		FileSize:         fileSize,
		DateTaken:        dateTaken,
		UploadedAt:       time.Now().UTC(),
		Orientation:      1, // Default: normal orientation
	}, nil
}

// sanitizeFilename removes path components and invalid characters
func sanitizeFilename(filename string) string {
	// Get just the filename, no path
	name := filepath.Base(filename)

	// Remove potentially dangerous characters
	replacer := strings.NewReplacer(
		"..", "",
		"/", "_",
		"\\", "_",
		":", "_",
		"*", "_",
		"?", "_",
		"\"", "_",
		"<", "_",
		">", "_",
		"|", "_",
	)

	return replacer.Replace(name)
}

// Errors
type PhotoError struct {
	Message string
}

func (e PhotoError) Error() string {
	return e.Message
}

var (
	ErrEmptyFilename    = PhotoError{"original filename cannot be empty"}
	ErrEmptyStoredPath  = PhotoError{"stored path cannot be empty"}
	ErrEmptyHash        = PhotoError{"file hash cannot be empty"}
	ErrInvalidFileSize  = PhotoError{"file size must be positive"}
	ErrPhotoNotFound    = PhotoError{"photo not found"}
	ErrDuplicatePhoto   = PhotoError{"photo already exists"}
	ErrInvalidExtension = PhotoError{"file extension not allowed"}
	ErrFileTooLarge     = PhotoError{"file size exceeds maximum allowed"}
	ErrPathTraversal    = PhotoError{"invalid path - path traversal detected"}
)
